### DISCUSSION TITLE
Negative xp support.

### DISCUSSION THREAD

**Discussion by WeylonSantana** [AscensionGameDev/Intersect-Engine#1492] [Ideas]
**Is your feature request related to a problem? Please describe.**
Change the give exp, to "set exp", and then accept positive and negative values to be added to or subtracted from the current xp.

Ps: Due to the #384 / #381 , currently is possible to remove xp from the player, just set a variable to a negative value and give xp to the player based on the value of the variable. (no errors, no bugs).

Which player reaches 0 exp. It is no longer possible to remove exp. (In case to make it return to the previous level.)

I am opening this feature request so that there may be (who knows) in the future the feature to subtract exp in the "normal" way too, in addition to the level being subtracted when the player reaches 0 exp when another attempt to subtract exp is executed. (in this case he would return to the previous level with 99% of the xp needed to return to the level he was in).

Another reason is that you probably didn't program this feature to subtract exp either, and then I would be using the engine in a way that I shouldn't (among many other things I use, in the way that I shouldn't)

**Describe the solution you'd like**
I ended up quoting the solution above

**Additional context**

**Comment by Arufonsu** [comment]
> **Is your feature request related to a problem? Please describe.**
> Change the give exp, to "set exp", and then accept positive and negative values to be added to or subtracted from the current xp.
> 
just an opinion
set would mean to **set** the exp value at x value, which is very different from "give".
perhaps a "take exp" could replace the ability to "give" minus x value of EXP.


> 
> Which player reaches 0 exp. It is no longer possible to remove exp. (In case to make it return to the previous level.)
> 
Sounds like a **Tibia** feature, pretty much the only MMO where i have seen the "decrease level" function, can't be impossible to add this feature, but consider that **decreasing a player's level** also means: decrease the gained stats from the previous level up, such as HP, MP, distributed points? may have to keep a track on all those values. Only few games follow this pattern out there, the common behaviour i have seen in games in general terms, is yes, some allow negative exp if u die around the edge of 0 exp, while others just stay stuck at 0 exp, but i haven't seen many using the decreasing level logic, anyway it does sounds like an awesome idea.

regards 😄 ~

**Comment by WeylonSantana** [comment]
@Arufonsu I never played Tibia, so I didn't understand much of what you mentioned.
I already played a lot of MMO where when dying was subtracted an XP from the player as punishment. MuOnline for example. If you mentioned tibia then it must be just another game that uses the same system.

With the command "define exp" it is possible to subtract xp as I mentioned. But that's when it reaches 0exp. It would be nice if when this value came to be negative, the player would return to the previous level with 99%exp, and if the level was 1, he would not return and stays with 0 exp. And yes I know that returning the level would mean decreasing HP, MP, distributed points. Intersect practically already has this system, where the player's attributes are updated when he changes from a higher level to a lower level and / or changes his class. All that's missing is this support.

**Comment by jcsnider** [comment]
Something else to consider is that if you have any items/spells granted on level up, or level up events that do stuff, those actions could not be undone while de-leveling.

That all being said, if someone makes a PR for this we will likely accept it.

**Comment by WeylonSantana** [comment]
@Cheshire92 This topic is a duplicate of this https://github.com/AscensionGameDev/Intersect-Engine/issues/722
This option was added in the engine, but reverted in the dev branch. https://github.com/AscensionGameDev/Intersect-Engine/pull/748
What to do? To close? Link the pr with this open?
It is currently "in progress"

**Comment by Cheshire92** [comment]
Is this a duplicate? I believe your request was to allow for it through events as well no?
As for the other issue, that PR still appears to be in the dev branch as I can see the options.

**Comment by WeylonSantana** [comment]
I understand the difference, I'll leave it open then

**Comment by WeylonSantana** [comment]
solved by me https://github.com/AscensionGameDev/Intersect-Engine/pull/2492
